source/slang/slang-ir-diff-jvp.cpp

```diff
@@ -9,373 +9,741 @@
+            differentialAssocTypeStructKey = parent->differentialAssocTypeStructKey;
+            isInterfaceAvailable = parent->isInterfaceAvailable;
+        }
+        else
+        {
+            differentiableInterfaceType = as<IRInterfaceType>(findDifferentiableInterface());
+            if (differentiableInterfaceType)
+            {
+                differentialAssocTypeStructKey = findDifferentialTypeStructKey();
+
+                if (differentialAssocTypeStructKey)
+                    isInterfaceAvailable = true;
+            }
+        }
+
+        if (isInterfaceAvailable)
+        {
+            // Load all witness tables corresponding to the IDifferentiable interface.
+            loadWitnessTablesForInterface(differentiableInterfaceType);
+        }
+    }
+
+    DifferentiableTypeConformanceContext(IRInst* inst) :
+        DifferentiableTypeConformanceContext(nullptr, inst)
+    {}
+
+    // Lookup a witness table for the concreteType. One should exist if concreteType
+    // inherits (successfully) from IDifferentiable.
+    // 
+    IRInst* lookUpConformanceForType(IRInst* type)
+    {
+        SLANG_ASSERT(isInterfaceAvailable);
+
+        if (witnessTableMap.ContainsKey(type))
+            return witnessTableMap[type];
+        else if (parent)
+            return parent->lookUpConformanceForType(type);
+        else
+            return nullptr;
+    }
+    
+    // Lookup and return the 'Differential' type declared in the concrete type
+    // in order to conform to the IDifferentiable interface.
+    // Note that inside a generic block, this will be a witness table lookup instruction
+    // that gets resolved during the specialization pass.
+    // 
+    IRInst* getDifferentialForType(IRBuilder* builder, IRType* origType)
+    {
+        SLANG_ASSERT(isInterfaceAvailable);
+
+        if (auto conformance = lookUpConformanceForType(origType))
+        {
+            if (auto witnessTable = as<IRWitnessTable>(conformance))
+            {
+                for (auto entry : witnessTable->getEntries())
+                {
+                    if (entry->getRequirementKey() == differentialAssocTypeStructKey)
+                        return as<IRType>(entry->getSatisfyingVal());
+                }
+            }
+            else if (auto witnessTableParam = as<IRParam>(conformance))
+            {
+                return builder->emitLookupInterfaceMethodInst(
+                    builder->getTypeKind(),
+                    witnessTableParam,
+                    differentialAssocTypeStructKey);
+            }
+        }
+
+        return nullptr;
+    }
+
+    private:
+
+    IRInst* findDifferentiableInterface()
+    {
+        if (auto module = as<IRModuleInst>(inst))
+        {
+            for (auto globalInst : module->getGlobalInsts())
+            {
+                // TODO: This seems like a particularly dangerous way to look for an interface.
```

Comment:
We should turn `IDifferentiable` as another `__intrinsic_type` so we can map it directly  into an IR type.
We should also add something like `__intrinsic_op` on the `associatedtype` declaration in `IDifferentiable` interface to turn all `lookupWitnessTable` insts on the associatedtype into an `IRGetDifferentiableType` inst during lower-to-ir.